sopel/bot.py

```diff
@@ -386,32 +388,84 @@ def register(self, callables, jobs, shutdowns, urls):
         :type shutdowns: :term:`iterable`
         :param urls: an iterable of functions to call when matched against a URL
         :type urls: :term:`iterable`
+
+        The ``callables`` argument contains a list of "callable objects", i.e.
+        object for witch :func:`callable` will return ``True``. They can be:
+
+        * a callable with rules (will match triggers with a regex pattern)
+        * a callable without rules (will match any triggers)
+        * a callable with commands
+        * a callable with nick commands
+
+        It is possible to have a callable with rules, commands, and nick
+        commands configured. It should not be possible to have a callable with
+        commands or nick commands but without rules.
         """
         # Append module's shutdown function to the bot's list of functions to
         # call on shutdown
         self.shutdown_methods += shutdowns
+        match_any = re.compile('.*')
         for callbl in callables:
-            if hasattr(callbl, 'rule'):
-                for rule in callbl.rule:
+            callable_name = getattr(callbl, "__name__", 'UNKNOWN')
+            rules = getattr(callbl, 'rule', [])
+            commands = getattr(callbl, 'commands', [])
+            nick_commands = getattr(callbl, 'nickname_commands', [])
+            is_rule_only = rules and not commands and not nick_commands
```

Comment:
As far as this logic goes, what about callables with `urls` (or `url_regex`, from the decorator)?

I'm glancing at this quickly while waiting for Travis to make sure my fix for the broken `translate` test comes through into this PR, so maybe I'm missing something.

Reply:
They are covered line 461 with the `for func in urls:`: URL Callbacks are not in the "callables" list. Same for jobs and shutdown functions, which are in separated lists.

Reply:
I knew I had to be missing something in my rush. :)